src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -503,8 +503,11 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
 
         private bool GetAwaitDisposeAsyncInfo(ref ForEachEnumeratorInfo.Builder builder, DiagnosticBag diagnostics)
         {
-            BoundExpression placeholder = new BoundAwaitableValuePlaceholder(_syntax.Expression,
-                this.GetWellKnownType(WellKnownType.System_Threading_Tasks_ValueTask, diagnostics, this._syntax));
+            var awaitableType = builder.DisposeMethod is null
+                ? this.GetWellKnownType(WellKnownType.System_Threading_Tasks_ValueTask, diagnostics, this._syntax)
+                : builder.DisposeMethod.ReturnType.TypeSymbol;
+
+            BoundExpression placeholder = new BoundAwaitableValuePlaceholder(_syntax.Expression, awaitableType);
```

Comment:
Nit: Why does the local change the type here of the `new` expression? It's not passed by `ref` hence couldn't see an obvious reason for this